### PR TITLE
Add AUC metric.

### PR DESCRIPTION
**Pull request by HarshTrivedi**
Adds AUC metric. This PR directly uses scikit-learn since it's already a dependency. If it's not preferred, I can change it with appropriate code.

### CODE REVIEW

File: allennlp/training/metrics/auc.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+from allennlp.training.metrics.metric import Metric
+
+
+@Metric.register("auc")
+class Auc(Metric):
+    """
+    The AUC Metric measures the area under the receiver-operating characteristic
+    (ROC) curve for binary classification problems.
+    """
+    def __init__(self, pos_label=1):
```
Comment (DeNeutoy) [inline allennlp/training/metrics/auc.py]: `positive_label`?

File: allennlp/training/metrics/auc.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+        self._all_predictions = torch.cat([self._all_predictions,
+                                           torch.masked_select(predictions, mask).float()], dim=0)
+        self._all_gold_labels = torch.cat([self._all_gold_labels,
+                                           torch.masked_select(gold_labels, mask).long()], dim=0)
+
+    def get_metric(self, reset: bool = False):
+        if self._all_gold_labels.shape[0] == 0:
+            return 0.5
+        fpr, tpr, _ = metrics.roc_curve(self._all_gold_labels.numpy(),
```
Comment (DeNeutoy) [inline allennlp/training/metrics/auc.py]: Full variable names please.

File: allennlp/tests/training/metrics/auc_test.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+from sklearn import metrics
+from numpy.testing import assert_almost_equal
+import pytest
+
+from allennlp.common.testing import AllenNlpTestCase
+from allennlp.training.metrics import Auc
+from allennlp.common.checks import ConfigurationError
+
+
+class AucTest(AllenNlpTestCase):
```
Comment (DeNeutoy) [inline allennlp/tests/training/metrics/auc_test.py]: Could you add a test that checks that the AUC metric doesn't break if you call `get_metrics` without ever having called the metric itself?

**Comment by HarshTrivedi** [comment]
@DeNeutoy, Thanks for the review! This is should be good to go now.

**Comment by HarshTrivedi** [comment]
@DeNeutoy, if this looks good, can you merge this?

**Comment by DeNeutoy** [comment]
Yep sorry!

File: allennlp/training/metrics/auc.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        if predictions.dim() != 1:
+            raise ConfigurationError("predictions must be one-dimensional, "
+                                     "but found tensor of shape: {}".format(predictions.size()))
+
+        unique_gold_labels = torch.unique(gold_labels)
+        if unique_gold_labels.numel() > 2:
+            raise ConfigurationError("AUC can be used for binary tasks only. gold_labels has {} unique labels, "
+                                     "expected at maximum 2.".format(unique_gold_labels.numel()))
+
+        gold_labels_is_binary = list(torch.sort(unique_gold_labels)[0].numpy()) == [0, 1]
```
Comment (generall) [inline allennlp/training/metrics/auc.py]: This may cause error is some batch contains only one value

Comment (Varal7) [reply allennlp/training/metrics/auc.py]: There is an attempt to fix this in #2775
